xcdat/temporal_avg.py

```diff
@@ -0,0 +1,986 @@
+
+# Type alias for supported time averaging modes.
+Mode = Literal["climatology", "timeseries", "departure"]
+MODES = get_args(Mode)
+
+# Type alias for supported grouping frequencies.
+Frequency = Literal["hour", "day", "month", "season", "year"]
+#: Tuple of supported grouping frequencies.
+FREQUENCIES = get_args(Frequency)
+
+# Configuration specific to the "season" frequency.
+SeasonConfig = TypedDict(
+    "SeasonConfig",
+    {
+        "dec_mode": Literal["DJF", "JFD"],
+        "drop_incomplete_djf": bool,
+        "custom_seasons": Optional[List[str]],
+    },
+    total=False,
+)
+SEASON_CONFIG_KEYS = ["dec_mode", "drop_incomplete_djf", "custom_seasons"]
+
+# Type alias representing xarray datetime accessor components.
+# https://xarray.pydata.org/en/stable/user-guide/time-series.html#datetime-components
+DateTimeComponent = Literal["hour", "day", "month", "season", "year"]
+
+# A dictionary mapping frequencies to xarray datetime accessor components. The
+# datetime components are used to create a pandas MultiIndex for grouping
+# operations. For the "season" frequency, "year and "month" must be included to
+# remove incomplete seasons and shift over Decembers a year for a "DJF" winter
+# season. Prior to grouping, "month" is removed from time series averaging and
+# "year" and "month" are removed from climatology.
+FREQ_TO_COMPONENTS = {
+    "timeseries": {
+        "year": ("year",),
+        "season": ("year", "season", "month"),
+        "month": ("year", "month"),
+        "day": ("year", "month", "day"),
+        "hour": ("year", "month", "day", "hour"),
+    },
+    "climatology": {
+        "season": ("year", "season", "month"),
+        "month": ("month",),
+        "day": ("month", "day"),
+    },
+    "departure": {
+        "season": ("year", "season", "month"),
+        "month": ("month",),
+        "day": ("month", "day"),
+    },
+}
+
+#: Dictionary mapping month integer to string.
+MONTH_INT_TO_STR: Dict[int, str] = {
+    1: "Jan",
+    2: "Feb",
+    3: "Mar",
+    4: "Apr",
+    5: "May",
+    6: "Jun",
+    7: "Jul",
+    8: "Aug",
+    9: "Sep",
+    10: "Oct",
+    11: "Nov",
+    12: "Dec",
+}
+
+
+@xr.register_dataset_accessor("temporal")
+class TemporalAverageAccessor:
+    def __init__(self, dataset: xr.Dataset):
+        try:
+            dataset.cf["T"]
+        except KeyError:
+            raise KeyError(
+                "This dataset does not have a time dimension, which is required for "
+                "using temporal operations in this accessor class."
+            )
+
+        # The Dataset object.
+        self._dataset: xr.Dataset = dataset
+
+        # The weights for time coordinates calculated based on a frequency.
+        self._weights: Optional[xr.DataArray] = None
+
+    def temporal_avg(
+        self,
+        mode: Mode,
+        freq: Frequency,
+        data_var: Optional[str] = None,
+        weighted: bool = True,
+        centered_time: bool = True,
```

Comment:
Consider renaming this `center_time`. `centered_time` (past tense) suggests the time has already been centered, but a value of `True` will result in `center_time()` being called.

Reply:
Thanks, I must have copied the suggestion from one our meetings without thinking about the tense.